<a href="https://colab.research.google.com/github/morreh89/TitanicEDA/blob/main/Naive_Bayes_Classifier_for__spam_detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction



> Spams are irrelevant or unsolicited (e.g. advertisements for products/web sites, make money fast schemes, chain letters, pornography) messages sent over the Internet, typically to a large number of users, for the purposes of advertising, phishing, spreading malware, etc


*   I have obtained an email dataset from the UCI Machine Learning Repository . The spam collection came from their postmaster and individuals who had filed spam whereas the non-spam e-mails came from filed work and personal e-mails, and hence the word 'george' and the area code '650' are indicators of non-spam. These are useful when constructing a personalized spam filter. One would either have to blind such non-spam indicators or get a very wide collection of non-spam to generate a general purpose spam filter. 






# Problem Statement



*   From the features seen in the dataset I am tasked to use Naive Bayes Classifier to decipher whether an email is a spam or non-spam.



# Evaluation Metrics



*  I am using the accuracy score and F1 score to evaluate the performance of our model.



# Reading Data

In [9]:
# importing libraries needed for that analysis
import pandas as pd
# allows us to read our files and organize them in tabular form
import numpy as np
# numpy allows us to work with multi dimensional arrays
import scipy
# scipy-a python ecosystem used for mathematics, machine learning 
import matplotlib.pyplot as plt
# matplotlib - a 2D library used to plot numpy computations
import seaborn as sns
# seaborn-a python visualization based on matplotlib, helps in drawing attractive statistical graphics
import sklearn
# sklearn-used for machine learning, 
%matplotlib inline
# for visualization 
import statsmodels.api as sm

# Machine Learning libraries

from sklearn.model_selection import train_test_split    # Used to split the data into the training and testing sets
from sklearn.preprocessing import StandardScaler        # Used to scale the data through normalization
from sklearn.preprocessing import Normalizer
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score              # Used as a metric for evaluation
from sklearn.metrics import classification_report

# Other libraries
import warnings
warnings.filterwarnings('ignore')                       # Used to ignore the warnings displayed by python 


In [10]:
#defining the three models of naive bayes that we are going to use.
gauss = GaussianNB()
multi = MultinomialNB()
bern = BernoulliNB()

In [11]:
# previewing the dataset

spam=pd.read_csv('spambase.csv', names=['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',
       'char_freq_%3B', 'char_freq_%28', 'char_freq_%5B', 'char_freq_%21',
       'char_freq_%24', 'char_freq_%23', 'capital_run_length_average',
       'capital_run_length_longest', 'capital_run_length_total', 'class'])
spam.take(np.random.permutation(len(spam))[:15])


word_freq_make  word_freq_address  ...  capital_run_length_total  class
713             0.00               0.00  ...                       194      1
3271            0.00               0.00  ...                        23      0
2546            0.00               0.00  ...                        37      0
4582            0.00               0.00  ...                        19      0
1073            1.61               0.00  ...                        64      1
3682            0.00               0.00  ...                         4      0
3700            0.00               0.00  ...                         6      0
925             0.00               0.75  ...                        93      1
3791            0.00               0.00  ...                        23      0
4019            0.00               0.00  ...                        31      0
1754            0.10               0.10  ...                      1042      1
3887            0.00               0.00  ...                       187      0
886             0.00               0.00  ...                       144      1
6               0.00               0.00  ...                       112      1
3048            0.00               0.00  ...                        25      0

[15 rows x 58 columns]



*   The last column of 'spambase.csv' denotes whether the e-mail was considered spam (1) or not (0), i.e. unsolicited commercial e-mail. Most of the attributes indicate whether a particular word or character was frequently occuring in the e-mail. The run-length attributes (55-57) measure the length of sequences of consecutive capital letters


---



> * ***Here is the description of the attributes***:



*   48 continuous real [0,100] attributes of type word_freq_WORD 
= percentage of words in the e-mail that match WORD, i.e. 100 * (number of times the word WORD appears in the e-mail) / total number of words in e-mail. A "word" in this case is any string of alphanumeric characters bounded by non-alphanumeric characters or end-of-string. 

* 6 continuous real [0,100] attributes of type char_freq_CHAR] 
= percentage of characters in the e-mail that match CHAR, i.e. 100 * (number of CHAR occurences) / total characters in e-mail 

* 1 continuous real [1,...] attribute of type capital_run_length_average 
= average length of uninterrupted sequences of capital letters 

* 1 continuous integer [1,...] attribute of type capital_run_length_longest 
= length of longest uninterrupted sequence of capital letters 

* 1 continuous integer [1,...] attribute of type capital_run_length_total 
= sum of length of uninterrupted sequences of capital letters 
= total number of capital letters in the e-mail 

* 1 nominal {0,1} class attribute of type spam 
= denotes whether the e-mail was considered spam (1) or not (0), i.e. unsolicited commercial e-mail. 












In [12]:
# describing the dataset
spam.describe().transpose()

count        mean  ...      75%        max
word_freq_make              4601.0    0.104553  ...    0.000      4.540
word_freq_address           4601.0    0.213015  ...    0.000     14.280
word_freq_all               4601.0    0.280656  ...    0.420      5.100
word_freq_3d                4601.0    0.065425  ...    0.000     42.810
word_freq_our               4601.0    0.312223  ...    0.380     10.000
word_freq_over              4601.0    0.095901  ...    0.000      5.880
word_freq_remove            4601.0    0.114208  ...    0.000      7.270
word_freq_internet          4601.0    0.105295  ...    0.000     11.110
word_freq_order             4601.0    0.090067  ...    0.000      5.260
word_freq_mail              4601.0    0.239413  ...    0.160     18.180
word_freq_receive           4601.0    0.059824  ...    0.000      2.610
word_freq_will              4601.0    0.541702  ...    0.800      9.670
word_freq_people            4601.0    0.093930  ...    0.000      5.550
word_freq_report            4601.0    0.058626  ...    0.000     10.000
word_freq_addresses         4601.0    0.049205  ...    0.000      4.410
word_freq_free              4601.0    0.248848  ...    0.100     20.000
word_freq_business          4601.0    0.142586  ...    0.000      7.140
word_freq_email             4601.0    0.184745  ...    0.000      9.090
word_freq_you               4601.0    1.662100  ...    2.640     18.750
word_freq_credit            4601.0    0.085577  ...    0.000     18.180
word_freq_your              4601.0    0.809761  ...    1.270     11.110
word_freq_font              4601.0    0.121202  ...    0.000     17.100
word_freq_000               4601.0    0.101645  ...    0.000      5.450
word_freq_money             4601.0    0.094269  ...    0.000     12.500
word_freq_hp                4601.0    0.549504  ...    0.000     20.830
word_freq_hpl               4601.0    0.265384  ...    0.000     16.660
word_freq_george            4601.0    0.767305  ...    0.000     33.330
word_freq_650               4601.0    0.124845  ...    0.000      9.090
word_freq_lab               4601.0    0.098915  ...    0.000     14.280
word_freq_labs              4601.0    0.102852  ...    0.000      5.880
word_freq_telnet            4601.0    0.064753  ...    0.000     12.500
word_freq_857               4601.0    0.047048  ...    0.000      4.760
word_freq_data              4601.0    0.097229  ...    0.000     18.180
word_freq_415               4601.0    0.047835  ...    0.000      4.760
word_freq_85                4601.0    0.105412  ...    0.000     20.000
word_freq_technology        4601.0    0.097477  ...    0.000      7.690
word_freq_1999              4601.0    0.136953  ...    0.000      6.890
word_freq_parts             4601.0    0.013201  ...    0.000      8.330
word_freq_pm                4601.0    0.078629  ...    0.000     11.110
word_freq_direct            4601.0    0.064834  ...    0.000      4.760
word_freq_cs                4601.0    0.043667  ...    0.000      7.140
word_freq_meeting           4601.0    0.132339  ...    0.000     14.280
word_freq_original          4601.0    0.046099  ...    0.000      3.570
word_freq_project           4601.0    0.079196  ...    0.000     20.000
word_freq_re                4601.0    0.301224  ...    0.110     21.420
word_freq_edu               4601.0    0.179824  ...    0.000     22.050
word_freq_table             4601.0    0.005444  ...    0.000      2.170
word_freq_conference        4601.0    0.031869  ...    0.000     10.000
char_freq_%3B               4601.0    0.038575  ...    0.000      4.385
char_freq_%28               4601.0    0.139030  ...    0.188      9.752
char_freq_%5B               4601.0    0.016976  ...    0.000      4.081
char_freq_%21               4601.0    0.269071  ...    0.315     32.478
char_freq_%24               4601.0    0.075811  ...    0.052      6.003
char_freq_%23               4601.0    0.044238  ...    0.000     19.829
capital_run_length_average  4601.0    5.191515  ...    3.706   1102.5

In [13]:
# checking on the dataset

spam.dtypes

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 float64
word_freq_ge



*   The data that we have been provided with is purely numerical in nature



In [14]:
# description of the data
spam.describe().transpose()

count        mean  ...      75%        max
word_freq_make              4601.0    0.104553  ...    0.000      4.540
word_freq_address           4601.0    0.213015  ...    0.000     14.280
word_freq_all               4601.0    0.280656  ...    0.420      5.100
word_freq_3d                4601.0    0.065425  ...    0.000     42.810
word_freq_our               4601.0    0.312223  ...    0.380     10.000
word_freq_over              4601.0    0.095901  ...    0.000      5.880
word_freq_remove            4601.0    0.114208  ...    0.000      7.270
word_freq_internet          4601.0    0.105295  ...    0.000     11.110
word_freq_order             4601.0    0.090067  ...    0.000      5.260
word_freq_mail              4601.0    0.239413  ...    0.160     18.180
word_freq_receive           4601.0    0.059824  ...    0.000      2.610
word_freq_will              4601.0    0.541702  ...    0.800      9.670
word_freq_people            4601.0    0.093930  ...    0.000      5.550
word_freq_report            4601.0    0.058626  ...    0.000     10.000
word_freq_addresses         4601.0    0.049205  ...    0.000      4.410
word_freq_free              4601.0    0.248848  ...    0.100     20.000
word_freq_business          4601.0    0.142586  ...    0.000      7.140
word_freq_email             4601.0    0.184745  ...    0.000      9.090
word_freq_you               4601.0    1.662100  ...    2.640     18.750
word_freq_credit            4601.0    0.085577  ...    0.000     18.180
word_freq_your              4601.0    0.809761  ...    1.270     11.110
word_freq_font              4601.0    0.121202  ...    0.000     17.100
word_freq_000               4601.0    0.101645  ...    0.000      5.450
word_freq_money             4601.0    0.094269  ...    0.000     12.500
word_freq_hp                4601.0    0.549504  ...    0.000     20.830
word_freq_hpl               4601.0    0.265384  ...    0.000     16.660
word_freq_george            4601.0    0.767305  ...    0.000     33.330
word_freq_650               4601.0    0.124845  ...    0.000      9.090
word_freq_lab               4601.0    0.098915  ...    0.000     14.280
word_freq_labs              4601.0    0.102852  ...    0.000      5.880
word_freq_telnet            4601.0    0.064753  ...    0.000     12.500
word_freq_857               4601.0    0.047048  ...    0.000      4.760
word_freq_data              4601.0    0.097229  ...    0.000     18.180
word_freq_415               4601.0    0.047835  ...    0.000      4.760
word_freq_85                4601.0    0.105412  ...    0.000     20.000
word_freq_technology        4601.0    0.097477  ...    0.000      7.690
word_freq_1999              4601.0    0.136953  ...    0.000      6.890
word_freq_parts             4601.0    0.013201  ...    0.000      8.330
word_freq_pm                4601.0    0.078629  ...    0.000     11.110
word_freq_direct            4601.0    0.064834  ...    0.000      4.760
word_freq_cs                4601.0    0.043667  ...    0.000      7.140
word_freq_meeting           4601.0    0.132339  ...    0.000     14.280
word_freq_original          4601.0    0.046099  ...    0.000      3.570
word_freq_project           4601.0    0.079196  ...    0.000     20.000
word_freq_re                4601.0    0.301224  ...    0.110     21.420
word_freq_edu               4601.0    0.179824  ...    0.000     22.050
word_freq_table             4601.0    0.005444  ...    0.000      2.170
word_freq_conference        4601.0    0.031869  ...    0.000     10.000
char_freq_%3B               4601.0    0.038575  ...    0.000      4.385
char_freq_%28               4601.0    0.139030  ...    0.188      9.752
char_freq_%5B               4601.0    0.016976  ...    0.000      4.081
char_freq_%21               4601.0    0.269071  ...    0.315     32.478
char_freq_%24               4601.0    0.075811  ...    0.052      6.003
char_freq_%23               4601.0    0.044238  ...    0.000     19.829
capital_run_length_average  4601.0    5.191515  ...    3.706   1102.5

In [15]:
# obtain dataset's info

spam.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 



*   From this we can see the dataset has 4601 entries and 58 columns


In [16]:
# Select duplicate rows except first occurrence based on all columns
duplicateRows = spam[spam.duplicated()]
 
print("Duplicated Rows except first occurrence based on all columns are :")
print(duplicateRows)

Duplicated Rows except first occurrence based on all columns are :
      word_freq_make  word_freq_address  ...  capital_run_length_total  class
26               0.0                0.0  ...                        82      1
103              0.0                0.0  ...                       292      1
104              0.0                0.0  ...                        66      1
105              0.0                0.0  ...                       292      1
106              0.0                0.0  ...                        66      1
...              ...                ...  ...                       ...    ...
4439             0.0                0.0  ...                       282      0
4441             0.0                0.0  ...                       282      0
4537             0.0                0.0  ...                         6      0
4541             0.0                0.0  ...                         2      0
4550             0.0                0.0  ...                         4     

In [17]:
spam.duplicated().sum()

391

In [18]:
# dropping the duplicated data

spam.drop_duplicates(inplace=True)
spam.duplicated().sum()

0



*   The features in the dataset are clearly not normally distributed and we therefore need to normalize the data



In [19]:
corr = spam.corr()
corr

word_freq_make  ...     class
word_freq_make                    1.000000  ...  0.129194
word_freq_address                 0.034012  ...  0.100651
word_freq_all                     0.063207  ...  0.172350
word_freq_3d                      0.005388  ...  0.056383
word_freq_our                     0.021740  ...  0.230074
word_freq_over                    0.055067  ...  0.212311
word_freq_remove                  0.011061  ...  0.334456
word_freq_internet               -0.004370  ...  0.200666
word_freq_order                   0.103842  ...  0.221454
word_freq_mail                    0.040829  ...  0.131686
word_freq_receive                 0.063031  ...  0.272451
word_freq_will                    0.075911  ... -0.014606
word_freq_people                  0.061985  ...  0.118024
word_freq_report                  0.032263  ...  0.053534
word_freq_addresses               0.022454  ...  0.183849
word_freq_free                    0.065443  ...  0.279261
word_freq_business                0.081471  ...  0.260773
word_freq_email                   0.051571  ...  0.188321
word_freq_you                     0.120563  ...  0.256242
word_freq_credit                  0.020924  ...  0.186074
word_freq_your                    0.121458  ...  0.394392
word_freq_font                   -0.025574  ...  0.093727
word_freq_000                     0.126478  ...  0.325716
word_freq_money                   0.205689  ...  0.205045
word_freq_hp                     -0.075054  ... -0.270024
word_freq_hpl                    -0.063467  ... -0.242411
word_freq_george                 -0.048086  ... -0.168925
word_freq_650                    -0.050054  ... -0.164727
word_freq_lab                    -0.041541  ... -0.137913
word_freq_labs                   -0.054241  ... -0.180638
word_freq_telnet                 -0.038899  ... -0.130764
word_freq_857                    -0.031634  ... -0.120141
word_freq_data                   -0.042715  ... -0.124025
word_freq_415                    -0.026322  ... -0.118373
word_freq_85                     -0.046083  ... -0.154754
word_freq_technology             -0.057296  ... -0.145356
word_freq_1999                   -0.058771  ... -0.201597
word_freq_parts                  -0.008626  ... -0.034275
word_freq_pm                     -0.011199  ... -0.129383
word_freq_direct                 -0.035467  ... -0.061989
word_freq_cs                     -0.005280  ... -0.105920
word_freq_meeting                -0.026546  ... -0.142813
word_freq_original               -0.025603  ... -0.146063
word_freq_project                -0.022980  ... -0.099406
word_freq_re                     -0.037848  ... -0.152895
word_freq_edu                    -0.034455  ... -0.152805
word_freq_table                  -0.000950  ... -0.046481
word_freq_conference             -0.018904  ... -0.089188
char_freq_%3B                    -0.027248  ... -0.067125
char_freq_%28                    -0.019636  ... -0.106572
char_freq_%5B                    -0.034784  ... -0.070397
char_freq_%21                     0.059184  ...  0.233812
char_freq_%24                     0.101957  ...  0.326603
char_freq_%23                    -0.009149  ...  0.070315
capital_run_length_average        0.044783  ...  0.110274
capital_run_length_longest        0.059128  ...  0.203657
capital_run_length_total          0.084127  ...  0.231671
class                             0.129194  ...  1.000000

[58 rows x 58 columns]

In [ ]:
# Use of a pairplot to check for correlation between variables.

sns.pairplot(spam, kind = 'scatter', dropna = False)
plt.tight_layout
plt.show()

* An important decision point when working with a sample of data is whether to use parametric or nonparametric statistical methods.
* Parametric statistical methods assume that the data has a known and specific distribution, often a Gaussian distribution. If a data sample is not Gaussian, then the assumptions of parametric statistical tests are violated and nonparametric statistical methods must be used.



*  There are three tests that can be used to test the distribution of data i.e. D’Agostino’s K^2 Test,Anderson-Darling Test and Shapiro-Wilk Test
*   We will use the Shapiro test to test for normality.
* It evaluates a data sample and quantifies how likely it is that the data was drawn from a Gaussian distribution, named for Samuel Shapiro and Martin Wilk.
* It's believed to be a reliable test of normality, although there is some suggestion that the test may be suitable for smaller samples of data, e.g. thousands of observations or fewer.

In [ ]:
# checking on the distribution frequency of the dataset before modelling
# creating a loop to iterate through all the columns systematically.

# importing the shapiro function
from scipy.stats import shapiro 

for i in spam.columns:

  stat, p = shapiro(spam[i])             # testing for normality
  print('Statistics=%.3f, p=%.3f' % (stat, p))

# interpreting the results
alpha = 0.05
if p > alpha:
	print('Sample looks Gaussian')
else:
	print('Sample does not look Gaussian')

## Baseline model

In [ ]:
# Splitting the data into dependent and independent sets.

X = spam.drop('class', axis = 1)
y = spam['class']


In [ ]:
# Normalizing the features 

norm = Normalizer().fit(X)

### Gaussian model

In [ ]:
# Training the model.
# Splitting the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 55)

gauss = GaussianNB().fit(X_train, y_train)

# Predicting
y_pred = gauss.predict(X_test)

# Evaluating the predictions made by the model

# 1. Using Classification report
print(classification_report(y_test, y_pred))


# 2. Using  the accuracy score
print(confusion_matrix(y_test, y_pred))
print('The accuracy :',accuracy_score(y_pred,y_test))

### Bernoulli model

In [ ]:
# Training the model.
# Splitting the data into train and test sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 55)


bern = BernoulliNB().fit(X_train, y_train)

# Predicting
y_pred = bern.predict(X_test)

# Evaluating the predictions made by the model

# 1. Using Classification report
print(classification_report(y_test, y_pred))


# 2. Using  the accuracy score
print(confusion_matrix(y_test, y_pred))
print('The accuracy :',accuracy_score(y_pred,y_test))

### Multinomial NB Model

In [ ]:
# Training the model.
# Splitting the data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 55)


multi = MultinomialNB().fit(X_train, y_train)

# Predicting
y_pred = multi.predict(X_test)

# Evaluating the predictions made by the model

# 1. Using Classification report
print(classification_report(y_test, y_pred))


# 2. Using  the accuracy score
print(confusion_matrix(y_test, y_pred))
print('The accuracy :',accuracy_score(y_pred,y_test))

## Conclusion



*   After having normalized the data we see that the Bernoulli classifier has the best accuracy score at 89% and hence it is recommended for this dataset

